GiniVision/Classes/Camera.swift

```diff
@@ -55,89 +77,101 @@ internal class Camera {
+        var videoDevice: AVCaptureDevice? {
+            let devices = AVCaptureDevice.devices(withMediaType: AVMediaTypeVideo).filter {
+                ($0 as? AVCaptureDevice)?.position == .back
+            }
             guard let device = devices.first as? AVCaptureDevice else { return nil }
             return device
         }
         
-        let videoDevice = deviceWithMediaType(AVMediaTypeVideo, preferringPosition: .back)
         do {
             self.videoDeviceInput = try AVCaptureDeviceInput(device: videoDevice)
         } catch let error as NSError {
-            print("Could not create video device input \(error)")
             if error.code == AVError.Code.applicationIsNotAuthorizedToUseDevice.rawValue {
                 throw CameraError.notAuthorizedToUseDevice
             } else {
                 throw CameraError.unknown
             }
         }
-        
-        self.session.beginConfiguration()
+    }
+    
+    fileprivate func setupInput() {
         // Specify that we are capturing a photo, this will reset the format to be 4:3
         self.session.sessionPreset = AVCaptureSessionPresetPhoto
         if self.session.canAddInput(self.videoDeviceInput) {
             self.session.addInput(self.videoDeviceInput)
         } else {
             print("Could not add video device input to the session")
         }
-        
+    }
+    
+    fileprivate func setupPhotoCaptureOutput() {
         let output = AVCaptureStillImageOutput()
+        
         if self.session.canAddOutput(output) {
-            output.outputSettings = [ AVVideoCodecKey: AVVideoCodecJPEG ];
+            output.outputSettings = [AVVideoCodecKey: AVVideoCodecJPEG]
             self.session.addOutput(output)
             self.stillImageOutput = output
         } else {
             print("Could not add still image output to the session")
         }
+    }
+    
+    fileprivate func setupQRScanningOutput() {
+        let qrOutput = AVCaptureMetadataOutput()
+        
+        if self.session.canAddOutput(qrOutput) {
+            self.session.addOutput(qrOutput)
+            
+            qrOutput.setMetadataObjectsDelegate(self, queue: sessionQueue)
+            qrOutput.metadataObjectTypes = [AVMetadataObjectTypeQRCode]
+        } else {
+            print("Could not add metadata output to the session")
+        }
+    }
+}
+
+// MARK: AVCaptureMetadataOutputObjectsDelegate
+
+extension Camera: AVCaptureMetadataOutputObjectsDelegate {
+    func captureOutput(_ output: AVCaptureOutput!,
+                       didOutputMetadataObjects metadataObjects: [Any]!,
+                       from connection: AVCaptureConnection!) {
+        if metadataObjects.isEmpty {
+            return
+        }
         
-        self.session.commitConfiguration()
+        if let metadataObj = metadataObjects[0] as? AVMetadataMachineReadableCodeObject,
```

Comment:
Instead of checking if the array is empty and then getting the first object, can't you do a `if let metadataObjects.first`?

Reply:
I kept it like this since I will probably change it in the QR processing branch.